p2p/gossip_tracer.go

```diff
@@ -0,0 +1,115 @@
+package p2p
+
+import (
+	"strings"
+
+	"github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/peerstore"
+	"github.com/libp2p/go-libp2p/core/protocol"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var _ = pubsub.RawTracer(gossipTracer{})
+
+const (
+	unknownAgent = "unknown"
+)
+
+// Metrics collection for gossipsub
+type gossipTracer struct {
+	host    host.Host
+	metrics *P2PMetrics
+}
+
+func NewGossipTracer(h host.Host) *gossipTracer {
+	return &gossipTracer{host: h, metrics: NewP2PMetrics()}
+}
+
+func (g gossipTracer) AddPeer(p peer.ID, proto protocol.ID) {
```

Comment:
is there any reason receiver (i.e. `g`) is not a pointer?

Reply:
If we use pointer receiver, then it would not implement the `RawTracer` interface